src/k8s/pkg/k8sd/features/cilium/gateway.go

```diff
@@ -125,3 +85,52 @@ func ApplyGateway(ctx context.Context, snap snap.Snap, gateway types.Gateway, ne
 		Message: enabledMsg,
 	}, nil
 }
+
+func disableGateway(ctx context.Context, snap snap.Snap, network types.Network) (types.FeatureStatus, error) {
+	m := snap.HelmClient()
+
+	// Delete our GatewayClass named ck-gateway
+	if _, err := m.Apply(ctx, chartGatewayClass, helm.StateDeleted, nil); err != nil {
+		err = fmt.Errorf("failed to delete Gateway API GatewayClass: %w", err)
+		return types.FeatureStatus{
+			Enabled: false,
+			Version: ciliumAgentImageTag,
+			Message: fmt.Sprintf(gatewayDeleteFailedMsgTmpl, err),
+		}, err
+	}
+
+	changed, err := m.Apply(ctx, chartCilium, helm.StateUpgradeOnlyOrDeleted(network.GetEnabled()), map[string]any{"gatewayAPI": map[string]any{"enabled": false}})
+	if err != nil {
+		err = fmt.Errorf("failed to delete Gateway API cilium configuration: %w", err)
+		return types.FeatureStatus{
+			Enabled: false,
+			Version: ciliumAgentImageTag,
+			Message: fmt.Sprintf(gatewayDeleteFailedMsgTmpl, err),
+		}, err
+	}
+
+	// Remove Gateway CRDs if the Gateway feature is disabled.
+	// This is done after the Cilium update as cilium requires the CRDs to be present for cleanups.
+	if _, err := m.Apply(ctx, chartGateway, helm.StateDeleted, nil); err != nil {
+		err = fmt.Errorf("failed to delete Gateway API CRDs: %w", err)
+		return types.FeatureStatus{
+			Enabled: false,
+			Version: ciliumAgentImageTag,
+			Message: fmt.Sprintf(gatewayDeleteFailedMsgTmpl, err),
+		}, err
+	}
+
+	if !changed {
+		return types.FeatureStatus{
+			Enabled: false,
+			Version: ciliumAgentImageTag,
+			Message: disabledMsg,
+		}, nil
+	}
+
+	return types.FeatureStatus{
+		Enabled: false,
+		Version: ciliumAgentImageTag,
+		Message: disabledMsg,
+	}, nil
```

Comment:
this looks wrong. if we changed cilium, we need to rollout restart. if not, just return early

Reply:
True, that slipped through. Thanks!